File: problems/chapter14/nicetak/arc005_3.jl
using DataStructures

function main()
    H, W = parse.(Int, split(readline()))
    mp = [readline() for _ = 1:H]

    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    
    xₛ, yₛ, xₑ, yₑ = -1, -1, -1, -1
    for i = 1:H, j = 1:W
        if mp[i][j] == 's'
            xₛ, yₛ = i, j
        elseif mp[i][j] == 'g'
            xₑ, yₑ = i, j
        end
    end

    que = Deque{Tuple{Int, Int}}()
    pushfirst!(que, (xₛ, yₛ))
    dists = 10^9 * ones(Int, H, W)
    dists[xₛ, yₛ] = 0

    while !isempty(que)
        x, y = popfirst!(que)

        for d = 1:4
            x′ = x + dx[d]
            y′ = y + dy[d]

            if 1 ≤ x′ ≤ H && 1 ≤ y′ ≤ W
                if mp[x′][y′] == '#'
                    if dists[x′, y′] > dists[x, y] + 1
                        dists[x′, y′] = dists[x, y] + 1
                        push!(que, (x′, y′))
                    end
                else
                    if dists[x′, y′] > dists[x, y]
                        dists[x′, y′] = dists[x, y]
                        pushfirst!(que, (x′, y′))
                    end
                end
            end
        end
    end


    if dists[xₑ, yₑ] ≤ 2
        println("YES")
    else
        println("NO")
    end
        
end

main()

File: problems/chapter14/nicetak/arc005_3.py
from collections import deque

def main():
    H, W = map(int, input().split())
    mp = [input() for _ in range(H)]

    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]

    x_s, y_s, x_g, y_g = -1, -1, -1, -1
    for i in range(H):
        for j in range(W):
            if mp[i][j] == "s":
                x_s, y_s = i, j
            elif mp[i][j] == "g":
                x_g, y_g = i, j
    
    que = deque()
    que.appendleft((x_s, y_s))
    dists = [[10**9] * W for _ in range(H)]
    dists[x_s][y_s] = 0

    while len(que) > 0:
        x, y = que.popleft()

        for d in range(4):
            x_new = x + dx[d]
            y_new = y + dy[d]

            if (x_new >=0) & (x_new < H) & (y_new >= 0) & (y_new < W):
                if mp[x_new][y_new] == "#":
                    if dists[x_new][y_new] > dists[x][y] + 1:
                        dists[x_new][y_new] = dists[x][y] + 1
                        que.append((x_new, y_new))
                else:
                    if dists[x_new][y_new] > dists[x][y]:
                        dists[x_new][y_new] = dists[x][y]
                        que.appendleft((x_new, y_new))

    if dists[x_g][y_g] <= 2:
        print("YES")
    else:
        print("NO")


if __name__=='__main__':
    main()
